File: rust/edge/src/cp.rs
//! HTTPS client to the in-cluster control plane. Presents the edge's bearer
//! token; fetches the cert+key for an SNI with ETag revalidation. The token and
//! the returned private key only ever travel over this TLS connection.

use serde::Deserialize;

/// Outcome of a cert fetch.
pub enum CertFetch {
    /// `304` — the edge's cached copy is still current.
    Unchanged,
    /// `200` — new material (and its ETag, if the server sent one).
    Updated {
        chain_pem: Vec<u8>,
        key_pem: Vec<u8>,
        etag: Option<String>,
    },
}

#[derive(Deserialize)]
struct CertBody {
    chain_pem: String,
    key_pem: String,
}

/// Outcome of a WAF ruleset fetch.
pub enum WafFetch {
    /// `304` — the edge's cached ruleset is still current.
    Unchanged,
    /// `200` — new payload (generation, global YAML, zones, host→zone, + ETag).
    Updated {
        generation: u64,
        global_rules: String,
        zones: std::collections::HashMap<String, String>,
        host_zone_map: std::collections::HashMap<String, String>,
        etag: Option<String>,
    },
}

#[derive(Deserialize)]
struct WafBody {
    #[serde(default)]
    generation: u64,
    #[serde(default)]
    global_rules: String,
    #[serde(default)]
    zones: std::collections::HashMap<String, String>,
    #[serde(default)]
    host_zone_map: std::collections::HashMap<String, String>,
}

#[derive(Clone)]
pub struct CpClient {
    http: reqwest::Client,
    base: String,
    token: String,
}

impl CpClient {
    /// `base` is the control-plane URL (e.g. `https://controlplane:8443`).
    /// `ca_pem`, when present, is added as a trusted root (for a private CA).
    pub fn new(
        base: String,
        token: String,
        ca_pem: Option<Vec<u8>>,
    ) -> Result<Self, reqwest::Error> {
        let mut b = reqwest::Client::builder();
        if let Some(ca) = ca_pem {
            if let Ok(cert) = reqwest::Certificate::from_pem(&ca) {
                b = b.add_root_certificate(cert);
            }
        }
        Ok(Self {
            http: b.build()?,
            base: base.trim_end_matches('/').to_string(),
            token,
        })
    }

    pub async fn fetch_cert(
        &self,
        sni: &str,
        current_etag: Option<&str>,
    ) -> Result<CertFetch, String> {
        let url = format!("{}/v1/certs", self.base);
        // `sni` goes in the query string (?sni=…); .query() percent-encodes it,
        // so a wildcard SNI like `*.acme.com` is transmitted safely.
        let mut req = self
            .http
            .get(&url)
            .query(&[("sni", sni)])
            .bearer_auth(&self.token);
        if let Some(etag) = current_etag {
            req = req.header(reqwest::header::IF_NONE_MATCH, etag);
        }
        let resp = req.send().await.map_err(|e| format!("request: {e}"))?;

        match resp.status() {
            reqwest::StatusCode::NOT_MODIFIED => Ok(CertFetch::Unchanged),
            reqwest::StatusCode::OK => {
                let etag = resp
                    .headers()
                    .get(reqwest::header::ETAG)
                    .and_then(|v| v.to_str().ok())
                    .map(str::to_string);
                let body: CertBody = resp.json().await.map_err(|e| format!("decode: {e}"))?;
                Ok(CertFetch::Updated {
                    chain_pem: body.chain_pem.into_bytes(),
                    key_pem: body.key_pem.into_bytes(),
                    etag,
                })
            }
            s => Err(format!("control plane returned {s} for sni {sni:?}")),
        }
    }

    /// Fetch the global WAF ruleset (raw YAML + generation), with ETag
    /// revalidation. Mirrors `fetch_cert`.
    pub async fn fetch_waf(&self, current_etag: Option<&str>) -> Result<WafFetch, String> {
        let url = format!("{}/v1/waf", self.base);
        let mut req = self.http.get(&url).bearer_auth(&self.token);
        if let Some(etag) = current_etag {
            req = req.header(reqwest::header::IF_NONE_MATCH, etag);
        }
        let resp = req.send().await.map_err(|e| format!("request: {e}"))?;

        match resp.status() {
            reqwest::StatusCode::NOT_MODIFIED => Ok(WafFetch::Unchanged),
            reqwest::StatusCode::OK => {
                let etag = resp
                    .headers()
                    .get(reqwest::header::ETAG)
                    .and_then(|v| v.to_str().ok())
                    .map(str::to_string);
                let body: WafBody = resp.json().await.map_err(|e| format!("decode: {e}"))?;
                Ok(WafFetch::Updated {
                    generation: body.generation,
                    global_rules: body.global_rules,
                    zones: body.zones,
                    host_zone_map: body.host_zone_map,
                    etag,
                })
            }
            s => Err(format!("control plane returned {s} for /v1/waf")),
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::sync::{Arc, Mutex};
    use tokio::io::{AsyncReadExt, AsyncWriteExt};
    use tokio::net::TcpListener;

    /// A one-shot HTTP/1.1 stub: serves a single request with `response`, and
    /// records the raw request head (request line + headers) for assertions.
    /// Returns the bound `http://127.0.0.1:port` base URL.
    async fn stub(response: &'static str, captured: Arc<Mutex<String>>) -> String {
        let listener = TcpListener::bind("127.0.0.1:0").await.unwrap();
        let addr = listener.local_addr().unwrap();
        tokio::spawn(async move {
            let (mut sock, _) = listener.accept().await.unwrap();
            let mut buf = [0u8; 4096];
            let n = sock.read(&mut buf).await.unwrap();
            *captured.lock().unwrap() = String::from_utf8_lossy(&buf[..n]).to_string();
            sock.write_all(response.as_bytes()).await.unwrap();
            sock.flush().await.unwrap();
        });
        format!("http://{addr}")
    }

    #[tokio::test]
    async fn fetch_200_parses_body_and_etag_and_sends_bearer() {
        let body = r#"{"chain_pem":"CHAIN","key_pem":"KEY"}"#;
        let resp = format!(
            "HTTP/1.1 200 OK\r\nETag: \"abc\"\r\nContent-Type: application/json\r\nContent-Length: {}\r\nConnection: close\r\n\r\n{}",
            body.len(),
            body
        );
        // leak to get 'static (test-only)
        let resp: &'static str = Box::leak(resp.into_boxed_str());
        let captured = Arc::new(Mutex::new(String::new()));
        let base = stub(resp, captured.clone()).await;

        let cp = CpClient::new(base, "tok-123".into(), None).unwrap();
        let got = cp.fetch_cert("acme.com", None).await.unwrap();
        match got {
            CertFetch::Updated {
                chain_pem,
                key_pem,
                etag,
            } => {
                assert_eq!(chain_pem, b"CHAIN");
                assert_eq!(key_pem, b"KEY");
                assert_eq!(etag.as_deref(), Some("\"abc\""));
            }
            _ => panic!("expected Updated"),
        }

        let req = captured.lock().unwrap().clone();
        assert!(req.starts_with("GET /v1/certs?sni=acme.com "), "path: {req}");
        assert!(
            req.contains("authorization: Bearer tok-123")
                || req.contains("Authorization: Bearer tok-123"),
            "missing bearer header: {req}"
        );
    }

    #[tokio::test]
    async fn fetch_304_is_unchanged_and_sends_if_none_match() {
        let resp = "HTTP/1.1 304 Not Modified\r\nConnection: close\r\n\r\n";
        let captured = Arc::new(Mutex::new(String::new()));
        let base = stub(resp, captured.clone()).await;

        let cp = CpClient::new(base, "tok".into(), None).unwrap();
        let got = cp.fetch_cert("acme.com", Some("\"v1\"")).await.unwrap();
        assert!(matches!(got, CertFetch::Unchanged));

        let req = captured.lock().unwrap().clone();
        assert!(
            req.to_lowercase().contains("if-none-match: \"v1\""),
            "missing INM: {req}"
        );
    }

    #[tokio::test]
    async fn fetch_403_is_error() {
        let resp = "HTTP/1.1 403 Forbidden\r\nConnection: close\r\n\r\n";
        let captured = Arc::new(Mutex::new(String::new()));
        let base = stub(resp, captured).await;
        let cp = CpClient::new(base, "tok".into(), None).unwrap();
        assert!(cp.fetch_cert("evil.com", None).await.is_err());
    }
}

File: edge/cp.go
package edge

import (
	"bytes"
	"crypto/tls"
	"crypto/x509"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"net/url"
	"strconv"
	"strings"
	"time"
)

// Response body caps: a compromised/buggy control plane shouldn't be able to
// OOM the edge during decode. Generous but finite (a cert+key is a few KiB; a
// WAF ruleset bundle for many zones can be larger).
const (
	maxCertBody  = 4 << 20  // 4 MiB
	maxWafBody   = 64 << 20 // 64 MiB
	maxPurgeBody = 16 << 20 // 16 MiB (a journal page of {seq,scope,host,uri} records)
)

// CpClient is the HTTPS client to the in-cluster control plane. It presents the
// edge's bearer token and revalidates cert/WAF material with ETags. The token
// and the returned private key only ever travel over this connection. Mirrors
// the Rust edge's CpClient: GET /v1/certs?sni= and GET /v1/waf, fail-static.
type CpClient struct {
	http  *http.Client
	base  string
	token string
}

// NewCpClient builds a client for base (e.g. https://controlplane:8443). caPEM,
// when non-empty, is added as a trusted root (for a private CA); an unparseable
// CA is ignored (system roots are used). A bad/empty base is reported.
func NewCpClient(base, token string, caPEM []byte) (*CpClient, error) {
	if _, err := url.Parse(base); err != nil {
		return nil, fmt.Errorf("edge: invalid control-plane endpoint %q: %w", base, err)
	}

	tlsConfig := &tls.Config{MinVersion: tls.VersionTLS12}
	if len(caPEM) > 0 {
		pool := x509.NewCertPool()
		if pool.AppendCertsFromPEM(caPEM) {
			tlsConfig.RootCAs = pool
		}
	}
	return &CpClient{
		http: &http.Client{
			Timeout:   30 * time.Second,
			Transport: &http.Transport{TLSClientConfig: tlsConfig},
		},
		base:  strings.TrimRight(base, "/"),
		token: token,
	}, nil
}

// CertFetch is the outcome of a cert fetch.
type CertFetch struct {
	// Unchanged is true on a 304 (the edge's cached copy is current).
	Unchanged bool
	// ChainPEM/KeyPEM/Etag carry new material on a 200.
	ChainPEM []byte
	KeyPEM   []byte
	Etag     string
	// CAID is the signer's target ca_id from the X-Parapet-CA-Id response header. It
	// is populated on EVERY status (200, 304, and even a 404) — the universal
	// force-re-mint signal that rides the edge's existing /v1/certs poll. "" means the
	// CP didn't advertise one (old CP / no signer); the edge fail-statics on "".
	CAID string
	// SignerFP is the CP's active signing fp from X-Parapet-Signing-Cert-Fp (every arm) —
	// the tuple's other half: an active=OLD→NEW flip changes THIS at an identical CAID,
	// so the edge must re-mint on the (CAID, SignerFP) tuple to obtain a NEW-signed leaf.
	SignerFP string
}

type certBody struct {
	ChainPEM string `json:"chain_pem"`
	KeyPEM   string `json:"key_pem"`
}

// FetchCert fetches the cert+key for sni with ETag revalidation. The sni is
// percent-encoded into ?sni= so a wildcard like *.acme.com transmits safely.
// Returns an error on any non-200/304 status (incl. 403/404) — the caller is
// fail-static. The ETag is carried verbatim (quotes included).
func (c *CpClient) FetchCert(sni, currentEtag string) (CertFetch, error) {
	u := c.base + "/v1/certs?sni=" + url.QueryEscape(sni)
	resp, err := c.do(u, currentEtag)
	if err != nil {
		return CertFetch{}, err
	}
	defer resp.Body.Close()

	// The X-Parapet-CA-Id header rides EVERY status. Read it once, up front, so the
	// force-re-mint signal is surfaced on the 304 (the steady-state carrier — almost
	// all responses are 304), the 200, AND the 404 (a missing-cert sni still learns
	// the CP target). Steady state is ~100% 304s, so the 304 arm is load-bearing.
	caID := resp.Header.Get("X-Parapet-CA-Id")
	signerFP := resp.Header.Get("X-Parapet-Signing-Cert-Fp")
	switch resp.StatusCode {
	case http.StatusNotModified:
		return CertFetch{Unchanged: true, CAID: caID, SignerFP: signerFP}, nil
	case http.StatusOK:
		var body certBody
		if err := json.NewDecoder(io.LimitReader(resp.Body, maxCertBody)).Decode(&body); err != nil {
			return CertFetch{}, fmt.Errorf("decode: %w", err)
		}
		return CertFetch{
			ChainPEM: []byte(body.ChainPEM),
			KeyPEM:   []byte(body.KeyPEM),
			Etag:     resp.Header.Get("ETag"),
			CAID:     caID,
			SignerFP: signerFP,
		}, nil
	default:
		// Surface the target with the error so the caller can still observe a flip
		// even when the per-sni cert is missing (404).
		return CertFetch{CAID: caID, SignerFP: signerFP}, fmt.Errorf("control plane returned %d for sni %q", resp.StatusCode, sni)
	}
}

// WafFetch is the outcome of a WAF ruleset fetch.
type WafFetch struct {
	// Unchanged is true on a 304.
	Unchanged bool
	// On a 200: the generation, global YAML, per-zone YAML, host->zone bindings,
	// and the ETag.
	Generation  uint64
	GlobalRules string
	Zones       map[string]string
	HostZoneMap map[string]string
	Etag        string
	CAID        string // signer target ca_id (secondary force-re-mint confirmer; 200 only)
	SignerFP    string // active signing fp (200 only)
}

type wafBody struct {
	Generation  uint64            `json:"generation"`
	GlobalRules string            `json:"global_rules"`
	Zones       map[string]string `json:"zones"`
	HostZoneMap map[string]string `json:"host_zone_map"`
	CAID        string            `json:"ca_id"`
	SigningFP   string            `json:"signing_cert_fp"`
}

// FetchWaf fetches the WAF payload (global YAML + zones + host->zone map) scoped
// to the edge's token, with ETag revalidation. A 404 ("waf distribution
// disabled") is treated like any other non-200/304: an error the caller handles
// fail-static (keeps last-good).
func (c *CpClient) FetchWaf(currentEtag string) (WafFetch, error) {
	resp, err := c.do(c.base+"/v1/waf", currentEtag)
	if err != nil {
		return WafFetch{}, err
	}
	defer resp.Body.Close()

	switch resp.StatusCode {
	case http.StatusNotModified:
		return WafFetch{Unchanged: true}, nil
	case http.StatusOK:
		var body wafBody
		if err := json.NewDecoder(io.LimitReader(resp.Body, maxWafBody)).Decode(&body); err != nil {
			return WafFetch{}, fmt.Errorf("decode: %w", err)
		}
		return WafFetch{
			Generation:  body.Generation,
			GlobalRules: body.GlobalRules,
			Zones:       body.Zones,
			HostZoneMap: body.HostZoneMap,
			Etag:        resp.Header.Get("ETag"),
			CAID:        body.CAID,
			SignerFP:    body.SigningFP,
		}, nil
	default:
		return WafFetch{}, fmt.Errorf("control plane returned %d for /v1/waf", resp.StatusCode)
	}
}

// PurgeFetch is the outcome of a cache-purge poll.
type PurgeFetch struct {
	// Disabled is true on a 404 ("purge distribution disabled" at the CP): a clean,
	// expected off-state the caller skips quietly (distinct from an unreachable CP).
	Disabled bool
	// FlushRequired is true when the edge's cursor fell behind the CP's retained
	// journal — the edge bumps its global epoch (lazy flush-all) and jumps to MaxSeq.
	FlushRequired bool
	// Entries are the new purges (seq > the requested cursor), already scoped to this
	// edge's allowed hosts by the CP.
	Entries []PurgeEntry
	// MaxSeq is the highest journal seq; the edge advances its cursor to it.
	MaxSeq uint64
}

type purgeBody struct {
	Entries       []PurgeEntry `json:"entries"`
	MaxSeq        uint64       `json:"max_seq"`
	FlushRequired bool         `json:"flush_required"`
}

// FetchPurges polls GET /v1/purges?since=<cursor> for cache-purge directives the
// edge hasn't applied. A 404 returns Disabled=true (err nil) so the caller can skip
// quietly when the CP isn't distributing purges; any other non-200 is an error the
// caller handles fail-static (keeps its applied epochs + cursor). No ETag: the
// since-cursor already makes the poll incremental and idempotent.
func (c *CpClient) FetchPurges(since uint64) (PurgeFetch, error) {
	u := c.base + "/v1/purges?since=" + strconv.FormatUint(since, 10)
	resp, err := c.do(u, "")
	if err != nil {
		return PurgeFetch{}, err
	}
	defer resp.Body.Close()

	switch resp.StatusCode {
	case http.StatusNotFound:
		return PurgeFetch{Disabled: true}, nil
	case http.StatusOK:
		var body purgeBody
		if err := json.NewDecoder(io.LimitReader(resp.Body, maxPurgeBody)).Decode(&body); err != nil {
			return PurgeFetch{}, fmt.Errorf("decode: %w", err)
		}
		return PurgeFetch{
			FlushRequired: body.FlushRequired,
			Entries:       body.Entries,
			MaxSeq:        body.MaxSeq,
		}, nil
	default:
		return PurgeFetch{}, fmt.Errorf("control plane returned %d for /v1/purges", resp.StatusCode)
	}
}

// EdgeCertFetch is the outcome of a data-plane client-cert issuance.
type EdgeCertFetch struct {
	ChainPEM []byte // leaf-first chain (leaf + edge CA); pair with the locally-held key
	NotAfter string // RFC3339, for renewal scheduling
	Serial   string
	CAID     string // signer ca_id, so the edge self-confirms convergence post-mint
	SignerFP string // active signing fp that minted this leaf
	// RetryAfter is the parsed Retry-After delay on a 429/503 (signer saturated),
	// returned ALONGSIDE the error so the coordinator can back off the whole fleet.
	RetryAfter time.Duration
}

type edgeCertReqBody struct {
	CSRPEM string `json:"csr_pem"`
}

type edgeCertRespBody struct {
	ChainPEM  string `json:"chain_pem"`
	NotAfter  string `json:"not_after"`
	Serial    string `json:"serial"`
	CAID      string `json:"ca_id"`
	SigningFP string `json:"signing_cert_fp"`
}

// FetchEdgeCert posts a CSR to POST /v1/edge-cert and returns the signed chain. The
// edge holds the matching private key locally; only the public-key CSR and the
// chain transit. Any non-200 is an error the caller handles fail-static (keeps the
// last-good in-memory cert).
func (c *CpClient) FetchEdgeCert(csrPEM []byte) (EdgeCertFetch, error) {
	reqBody, err := json.Marshal(edgeCertReqBody{CSRPEM: string(csrPEM)})
	if err != nil {
		return EdgeCertFetch{}, err
	}
	req, err := http.NewRequest(http.MethodPost, c.base+"/v1/edge-cert", bytes.NewReader(reqBody))
	if err != nil {
		return EdgeCertFetch{}, fmt.Errorf("request: %w", err)
	}
	req.Header.Set("Authorization", "Bearer "+c.token)
	req.Header.Set("Content-Type", "application/json")
	resp, err := c.http.Do(req)
	if err != nil {
		return EdgeCertFetch{}, fmt.Errorf("request: %w", err)
	}
	defer resp.Body.Close()
	if resp.StatusCode != http.StatusOK {
		_, _ = io.Copy(io.Discard, io.LimitReader(resp.Body, 4096))
		// On a saturated-signer shed (429/503), surface the Retry-After so the
		// coordinator honors fleet-aggregate backpressure (one token per edge does NOT
		// bound the aggregate). Returned with the error; the caller still fail-statics.
		var retryAfter time.Duration
		if resp.StatusCode == http.StatusTooManyRequests || resp.StatusCode == http.StatusServiceUnavailable {
			retryAfter = parseRetryAfter(resp.Header.Get("Retry-After"))
		}
		return EdgeCertFetch{RetryAfter: retryAfter}, fmt.Errorf("control plane returned %d for /v1/edge-cert", resp.StatusCode)
	}
	var body edgeCertRespBody
	if err := json.NewDecoder(io.LimitReader(resp.Body, maxCertBody)).Decode(&body); err != nil {
		return EdgeCertFetch{}, fmt.Errorf("decode: %w", err)
	}
	return EdgeCertFetch{
		ChainPEM: []byte(body.ChainPEM),
		NotAfter: body.NotAfter,
		Serial:   body.Serial,
		CAID:     body.CAID,
		SignerFP: body.SigningFP,
	}, nil
}

// parseRetryAfter parses an HTTP Retry-After header value: either delta-seconds
// (e.g. "5") or an HTTP-date. Returns 0 on empty/unparseable/negative.
func parseRetryAfter(v string) time.Duration {
	v = strings.TrimSpace(v)
	if v == "" {
		return 0
	}
	if secs, err := strconv.Atoi(v); err == nil {
		if secs < 0 {
			return 0
		}
		return time.Duration(secs) * time.Second
	}
	if t, err := http.ParseTime(v); err == nil {
		if d := time.Until(t); d > 0 {
			return d
		}
	}
	return 0
}

type trustBundleBody struct {
	CAID      string `json:"ca_id"`
	SigningFP string `json:"signing_cert_fp"`
}

// FetchTrustBundleSignal reads the (ca_id, active signer fp) tuple from the tokenless GET
// /v1/trust-bundle (the same public endpoint the core polls). It lets an mTLS edge observe
// a CA rotation OR an active=OLD→NEW flip even when it polls no per-domain cert and no WAF
// (serve-all, no traffic, WAF off) — it rides the edge-cert refresh tick, not a new loop.
// The bearer token is sent (the endpoint ignores it). Errors are the caller's to
// fail-static on.
func (c *CpClient) FetchTrustBundleSignal() (caID, signerFP string, err error) {
	resp, err := c.do(c.base+"/v1/trust-bundle", "")
	if err != nil {
		return "", "", err
	}
	defer resp.Body.Close()
	if resp.StatusCode != http.StatusOK {
		return "", "", fmt.Errorf("control plane returned %d for /v1/trust-bundle", resp.StatusCode)
	}
	var body trustBundleBody
	if err := json.NewDecoder(io.LimitReader(resp.Body, maxCertBody)).Decode(&body); err != nil {
		return "", "", fmt.Errorf("decode: %w", err)
	}
	return body.CAID, body.SigningFP, nil
}

// do issues an authorized GET, optionally with If-None-Match. The body must be
// closed by the caller.
func (c *CpClient) do(rawURL, ifNoneMatch string) (*http.Response, error) {
	req, err := http.NewRequest(http.MethodGet, rawURL, nil)
	if err != nil {
		return nil, fmt.Errorf("request: %w", err)
	}
	req.Header.Set("Authorization", "Bearer "+c.token)
	if ifNoneMatch != "" {
		req.Header.Set("If-None-Match", ifNoneMatch)
	}
	resp, err := c.http.Do(req)
	if err != nil {
		return nil, fmt.Errorf("request: %w", err)
	}
	// Drain non-200/304 bodies so the connection can be reused, but keep the
	// body open for the caller on 200.
	if resp.StatusCode != http.StatusOK && resp.StatusCode != http.StatusNotModified {
		_, _ = io.Copy(io.Discard, io.LimitReader(resp.Body, 4096))
	}
	return resp, nil
}
